### ISSUE TITLE
Low average precision value

### ISSUE THREAD

**Issue by vladimir-sevastyanov** [jfkirk/tensorrec#74]
I have a dataset with 12000 users and 500 items, and I used it to build a recommender model by different approaches. With user based collaborative filtering I have precision=0.1, with content based model precision=0.45. I expected to increase the precision value with TensorFlow by obvious reasons, but precision=0.0013, which does not make any sense.
I tried your basic_usage code, which generates ratings, user_features and item_features randomly by the function tensorrec.util.generate_dummy_data(), and I see precision=0.003. It makes sense because in this case we don't have any correlations between ratings and user-item features.
Please, clarify why real data and random numbers give the same level of precision.

**Comment by vladimir-sevastyanov** [comment]
I used your function get_book_crossing() as a prototype to load and convert the data in proper format--I just adjusted your code for my set of data.. So, I hope the data I use to fit the recommender model is valid.

**Comment by jfkirk** [comment]
Hey @vladimir-sevastyanov : What models are you using for the collaborative filters, and what methods are you using to calculate precision (both with the collaborative filter and with TensorRec)?

Can you paste a code snippet of the TensorRec model you're using?

**Comment by vladimir-sevastyanov** [comment]
This is my code:

import numpy as np
import tensorrec
from tensorrec.eval import fit_and_eval, eval_random_ranks_on_dataset

from dataLoad.datasets import get_data #, get_book_crossing


train_interactions, test_interactions, user_features, product_features = get_data(min_positive_score=1,
                                                                                  min_interactions_per_product=2, 
                                                                                  user_indicators=False, 
                                                                                  item_indicators=False,
                                                                                  cold_start_users=False, 
                                                                                  cold_start_items=False)

# Build the model with default parameters
model = tensorrec.TensorRec()

# Fit the model and get a result packet
fit_kwargs = {'epochs': 500, 'learning_rate': .01, 'n_sampled_items': 100, 'verbose': True}
result = fit_and_eval(model, user_features, product_features, train_interactions, test_interactions, fit_kwargs,
                              recall_k=10, precision_k=10, ndcg_k=10)

# Predict scores and ranks for all users and all items
predictions = model.predict(user_features=user_features,
                            item_features=product_features)
predicted_ranks = model.predict_rank(user_features=user_features,
                                     item_features=product_features)

# Calculate and print the recall at 10
r_at_k = tensorrec.eval.recall_at_k(predicted_ranks, train_interactions, k=10)
print(np.mean(r_at_k))

p_at_k = tensorrec.eval.precision_at_k(predicted_ranks, train_interactions, k=6)
print(np.mean(p_at_k))

**Comment by vladimir-sevastyanov** [comment]
**That's the code to convert the data:**
   
def get_data(min_positive_score=2, min_interactions_per_product=2, user_indicators=False, item_indicators=False,
                      cold_start_users=False, cold_start_items=False):


    ratings = []
    with open('C:/temp/RecommenderSystems/tensorrec-master/data/interactions.csv', 'rb') as ratings_file:
        ratings_lines = ratings_file.readlines()
        for line in ratings_lines:
            split_line = line.decode('iso-8859-1').replace('\r', '').replace('\n', '').split(',')

            # Throw out header
            if split_line[0] == 'TPID':
                continue

            tpid = int(split_line[0])
            offeringid = split_line[1]
            rating = int(split_line[2])
            rating = 1 if rating >= min_positive_score else 0

            # Skip 0 scores
            if rating == 0:
                continue

            ratings.append((tpid, offeringid, rating))

    product_metadata_raw = {}
    with open('C:/temp/RecommenderSystems/tensorrec-master/data/items.csv', 'rb') as products_file:
        products_lines = products_file.readlines()
        for line in products_lines:

            split_line = line.decode('iso-8859-1').replace('\r', '').replace('\n', '').split(',')

            # Throw out header
            if split_line[0] == 'OfferingID':
                continue

            # Store tuple of title, author, year, publisher, cover image URL
            offeringid = split_line[0]
            servicetype = split_line[1]
            technology = split_line[2]
            focusarea = split_line[3]

            # Convert year back to string for encoding
            product_metadata_raw[offeringid] = (servicetype, technology, focusarea)

    user_metadata_raw = {}
    with open('C:/temp/RecommenderSystems/tensorrec-master/data/Users.csv', 'rb') as users_file:
        users_lines = users_file.readlines()
        for line in users_lines:

            split_line = line.decode('iso-8859-1').replace('\r', '').replace('\n', '').split(',')

            # Throw out header
            if split_line[0] == 'TPID':
                continue

            tpid = split_line[0]
            country = split_line[1]
            companyname = split_line[2]

            # Convert country back to string for encoding
            user_metadata_raw[tpid] = (country, companyname)

    offeringid_counter = {}
    for _, offeringid, _ in ratings:
        offeringid_counter[offeringid] = offeringid_counter.get(offeringid, 0) + 1
    ratings = [r for r in ratings if offeringid_counter[r[1]] >= min_interactions_per_product]

    # Build map between users, offeringids, and rows/columns
    user_to_row_map = {}
    offeringid_to_column_map = {}
    for user, offeringid, _ in ratings:
        if user not in user_to_row_map:
            user_to_row_map[user] = len(user_to_row_map)
        if offeringid not in offeringid_to_column_map:
            offeringid_to_column_map[offeringid] = len(offeringid_to_column_map)
    row_to_user_map = {row: user for user, row in user_to_row_map.items()}
    column_to_offeringid_map = {col: offeringid for offeringid, col in offeringid_to_column_map.items()}

    interactions_raw = [(user_to_row_map[user], offeringid_to_column_map[offeringid], score) for user, offeringid, score in ratings]
    r, c, v = zip(*interactions_raw)
    interactions = sp.coo_matrix((v, (r, c)), dtype=np.float64)

    user_metadata = []
    for row in range(max(row_to_user_map) + 1):
        user = row_to_user_map[row]
        user_metadata.append(user_metadata_raw[str(user)])

    product_metadata = []
    for col in range(max(column_to_offeringid_map) + 1):
        offeringid = column_to_offeringid_map[col]
        try:
            product_metadata.append(product_metadata_raw[offeringid])
        except KeyError:
            product_metadata.append(('', '', '', '', ''))

    product_transformers = [
        ('servicetype_pipeline', Pipeline([
            ('servicetype_extractor', TupleExtractor(0)),
            ('servicetype_vectorizer', CountVectorizer(min_df=2)),
        ])),
        ('technology_pipeline', Pipeline([
            ('technology_extractor', TupleExtractor(1)),
            ('technology_vectorizer', PipelineLabelBinarizer(sparse_output=True)),
        ])),
        ('focusarea_pipeline', Pipeline([
            ('focusarea_extractor', TupleExtractor(2)),
            ('focusarea_vectorizer', PipelineLabelBinarizer(sparse_output=True)),
        ])),
    ]
    if item_indicators:
        product_transformers.append(('indicator', IndicatorFeature()))
    product_pipeline = FeatureUnion(product_transformers)
    product_features = product_pipeline.fit_transform(product_metadata)

    user_transformers = [
        ('country_pipeline', Pipeline([
            ('country_extractor', TupleExtractor(0)),
            ('country_vectorizer', CountVectorizer(min_df=2)),
        ])),
        ('companyname_pipeline', Pipeline([
            ('companyname_extractor', TupleExtractor(1)),
            ('companyname_vectorizer', PipelineLabelBinarizer(sparse_output=True)),
        ])),
    ]
    if user_indicators:
        user_transformers.append(('indicator', IndicatorFeature()))
    user_pipeline = FeatureUnion(user_transformers)
    user_features = user_pipeline.fit_transform(user_metadata)

    if cold_start_users:
        train_interactions, test_interactions = _split_interactions_cold_start_users(interactions)
    elif cold_start_items:
        train_interactions, test_interactions = _split_interactions_cold_start_items(interactions)
    else:
        train_interactions, test_interactions = _split_interactions_warm_start(interactions)

    return train_interactions, test_interactions, user_features, product_features

**Comment by jfkirk** [comment]
@vladimir-sevastyanov The data-conversion code looks okay to me -- I didn't notice any immediate issues.

I'm confused re: your precision concern. You said before that you had k=1000, but in this code k=6. When you received a precision value of .0013, what was your K value?

**Comment by vladimir-sevastyanov** [comment]
I experimented with different K values, but it does not impact the order of the precision value

**Comment by jfkirk** [comment]
K is deeply impactful to precision values. For example, here is Precision from LightFM:

```def precision_at_k(model, test_interactions, train_interactions=None,
                   k=10, user_features=None, item_features=None,
                   preserve_rows=False, num_threads=1, check_intersections=True):

    ranks = model.predict_rank(test_interactions,
                               train_interactions=train_interactions,
                               user_features=user_features,
                               item_features=item_features,
                               num_threads=num_threads,
                               check_intersections=check_intersections,
                               )

    ranks.data = np.less(ranks.data, k, ranks.data)

    precision = np.squeeze(np.array(ranks.sum(axis=1))) / k

    if not preserve_rows:
        precision = precision[test_interactions.getnnz(axis=1) > 0]

    return precision```

And this is precision in TensorRec:

```def precision_at_k(predicted_ranks, test_interactions, k=10, preserve_rows=False):

    positive_test_interactions = test_interactions > 0
    ranks = sp.csr_matrix(predicted_ranks * positive_test_interactions.A)
    ranks.data = np.less(ranks.data, (k + 1), ranks.data)

    precision = np.squeeze(np.array(ranks.sum(axis=1))).astype(float) / k

    if not preserve_rows:
        precision = precision[positive_test_interactions.getnnz(axis=1) > 0]

    return precision```

**Comment by vladimir-sevastyanov** [comment]
Can I use ratings from 1 to hundreds and thousands, or it must be binary: 0/1? 
In my case I have rating as the number of licenses purchased by a company (user). Some companies have a few licences, but a few of them have hundreds and even thousands. 
Is it possible that precision is calculated incorrectly because rating is not binary? Do I have to convert rating into a binary value?

**Comment by jfkirk** [comment]
Both (LightFM and TensorRec) precision calculations should tolerate any positive value just fine. If you try binarizing your data before passing it to the precision functions, and it changes the result, please let me know.

My immediate suspicion is that the low precision you saw (which was almost exactly 100x lower than the other precisions) is the result of a higher K value (1000, which is 100x higher). Can you confirm that you get wildly different precision values with the same K?

**Comment by vladimir-sevastyanov** [comment]
Yes, I confirm.
I just performed one more experiment: used the same sparse matrix to build TensorRec and LightFM models, and calculated precision for each model for the same k=10. 
TensorRec precision=0.000216
LightFM precision=0.23
I suggest you to compare results in debug mode, and find out the difference

**Comment by vladimir-sevastyanov** [comment]
**My code:**

import numpy as np
import tensorrec
from tensorrec.eval import fit_and_eval, eval_random_ranks_on_dataset

from dataLoad.datasets import get_data #, get_book_crossing


train_interactions, test_interactions, user_features, product_features = get_data(min_positive_score=1,
                                                                                  min_interactions_per_product=2, 
                                                                                  user_indicators=False, 
                                                                                  item_indicators=False,
                                                                                  cold_start_users=False, 
                                                                                  cold_start_items=False)

# Build the model_tr with default parameters
model_tr = tensorrec.TensorRec()

# Fit the model_tr and get a result packet
fit_kwargs = {'epochs': 500, 'learning_rate': .01, 'n_sampled_items': 100, 'verbose': True}
result = fit_and_eval(model_tr, user_features, product_features, train_interactions, test_interactions, fit_kwargs,
                              recall_k=10, precision_k=10, ndcg_k=10)

# Predict scores and ranks for all users and all items
predictions = model_tr.predict(user_features=user_features,
                            item_features=product_features)

# Calculate and print the recall at 10
p_at_k = tensorrec.eval.precision_at_k(predicted_ranks, train_interactions, k=10)
print(np.mean(p_at_k))

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
from lightfm import LightFM

model_lfm = LightFM(loss='warp')
model_lfm = LightFM(loss='bpr')
model_lfm.fit(train_interactions, epochs=50, num_threads=2)

from lightfm.evaluation import precision_at_k

print("Train precision: %.2f" % precision_at_k(model_lfm, train_interactions, k=10).mean())
print("Test precision: %.2f" % precision_at_k(model_lfm, test_interactions, k=10).mean())

**Comment by vladimir-sevastyanov** [comment]
When I used the suggested above variant
model_tr = tensorrec.TensorRec(loss_graph=tensorrec.loss_graphs.WMRBLossGraph())
I've got the error message "InvalidArgumentError (see above for traceback): 0-th value returned by pyfunc_0 is int32, but expects int64", which I described in the previous issue.

Default RMSE parameter does not generate the error:
model_tr = tensorrec.TensorRec()

**Comment by jfkirk** [comment]
hey @vladimir-sevastyanov -- I've been unable to reproduce the problem, and I got your LinkedIn note. I'm going to close this issue for now.
